src/support/topological_orders.cpp

```diff
@@ -134,11 +134,15 @@ TopologicalOrders& TopologicalOrders::operator++() {
 
 void TopologicalOrders::pushChoice(size_t choice) {
   choiceHeap.push_back(choice);
-  std::push_heap(choiceHeap.begin(), choiceHeap.end(), std::greater<size_t>{});
+  std::push_heap(choiceHeap.begin(), choiceHeap.end(), [&](size_t a, size_t b) {
+    return cmp(b, a);
+  });
 }
 
 size_t TopologicalOrders::popChoice() {
-  std::pop_heap(choiceHeap.begin(), choiceHeap.end(), std::greater<size_t>{});
+  std::pop_heap(choiceHeap.begin(), choiceHeap.end(), [&](size_t a, size_t b) {
+    return cmp(b, a);
+  });
```

Comment:
These indirect calls may add overhead. In ReorderGlobals I guess we have maybe on the order of 10K in real-world wasm files so maybe it isn't noticeable, but maybe it's worth measuring what happens if we template over cmp instead?

Reply:
Measuring the runtime of ReorderGlobals on a j2wasm binary with 30 samples each with templates or std::function, I get a mean time of 0.1457 seconds with templates and 0.1527 seconds with std::function, so about a 5% difference, and a p-value of 0.013 (assuming I applied my statistical tests correctly).

WDYT we should do based on that data?

Reply:
Is the template code more complex? If it's simple enough then it might be worthwhile. But for 5%, if it's complex or less readable then I guess not.

Reply:
The code is simple enough. The main downside is adding all that code to the header. I’ll experiment with splitting some of it into a separate implementation header.